Created first simple sharding test

blmartin opened the pull request:
Added test for load of shardtrie from json and split of shardtrie

blmartin committed:
Created first simple sharding test

Added test for load of shardtrie from json and split of shardtrie

blmartin committed:
Merged file from working branch

Added tests that were being written in the working branch.

matthewoliver commented on test/unit/common/test_shardtrie.py:
```diff
@@ -0,0 +1,312 @@
+        # Iterate starting at 'abc'
+        self.assertEqual(
+            list(trie.get_important_nodes('abc'))[0].full_key(),
+            'abc')
```
get_important_nodes iterates over dist and data nodes, so we should make sure a dist node it in the trie and show that it shows up.

matthewoliver replied:
Also, I added the limit stuff here cause originally I thought this would be where we'd use it. In the current interation (i'd have to check), because on container GET we grab CONTAINER_LISTING_LIMIT at a time, I think we are letting the container backend class deal with the limits for us, and we just keep track of how many are left.

TL;DR: I'm not sure we are actually using the limit stuff in shardtrie class... although maybe we should be.. just making a note cause one day we may decide to remove it on a clean up. But we'll decide on that later. :)

blmartin replied:
Added clause to split the trie and then get important nodes.
Verified and added to branch.

matthewoliver commented on test/unit/common/test_shardtrie.py:
```diff
@@ -0,0 +1,312 @@
+        # We cannot use the following line as is_empty checks that the key
+        #  is empty string, popping_iterator does not set the key to ''
+        # self.assertTrue(trie.is_empty())
+        self.assertEqual(trie.root.children, {})
```
Popping iterater was an attempt at being more memory efficient when getting.. it isn't used.. but might still be the right idea. This is where benchmarking will come in to play.

matthewoliver commented on test/unit/common/test_shardtrie.py:
```diff
@@ -0,0 +1,312 @@
+            trie.add('abd%d' % i)
+
+        large_tries = [x[2] for x in trie.get_large_subtries(15)]
+        self.assertEqual(large_tries, ['abd', 'abc', 'ab'])
```
This is the original way of getting the largest subtries.. it works so I say we leave it in for now. This uses way too much memory when dealing with _very_ large containers, and hence gave birth to the counting trie technique.

matthewoliver commented:
On an initial first pass, this is awesome! Its a great start. Sure there are more edge cases in some of the tests to check.. but we can add to them as we need them.

I wont merge yet as I want you to see the comments :)

Once you have I'm happy to merge this pull request! Awesome work!

matthewoliver commented on test/unit/common/test_shardtrie.py:
```diff
@@ -0,0 +1,312 @@
+        """Test getting the best candidate found by the counting trie"""
+        trie = self._setup_trie(group_count=2)
+
+        self.assertEqual(trie.get_best_candidates(), (2, ['ab']))
```
There is also a case where the counting trie may find more then 1 candidate at the lowest level in the trie, so we should test that too, maybe after this assert (line 279) we add:

  keys = ['a', 'abc', 'abd', 'cba', 'cbb']
  trie = self._setup_trie(keys=keys, group_count=2)
  self.assertEqual(trie.get_best_candidates(), (2, ['ab', 'cb']))

blmartin replied:
Added code to test case and verified. Thanks!

blmartin committed:
Ammended test for get_important_nodes

Added check for get_important_nodes when a node in the trie is
 distributed.

blmartin committed:
Extended test for CountingTrie find candidate

Added check in test for the case that the CountingTrie finds the
 more than one candidate. Addition created by matthewoliverau.

blmartin committed:
Pep8 compliance for ShardTrie test file

Fixed whitespace issues and '!= None' to 'is not None'

blmartin commented:
I have gone through the comments and added the changes you suggested. The greater picture is all starting to come into focus! Thanks for looking over the code.

blmartin commented:
Ah quick notice though, if merged now, some of the tests will fail due to some bug fixes not being present. I will go ahead and make pull requests for them now

matthewoliver commented:
Cool, will merge this now, knowing that some tests may fail :)
